A better description for "Requires-Python" metadata field (#1274)
pypa/packaging.python.org: source/specifications/core-metadata.rst

.. versionadded:: 1.2

This field specifies the Python version(s) that the distribution is
compatible with. Installation tools may look at this when
picking which version of a project to install.

The value must be in the format specified in :doc:`version-specifiers`.

For example, if a distribution uses `f-strings <https://docs.python.org/3/whatsnew/3.6.html#pep-498-formatted-string-literals>`_
then it may prevent installation on Python < 3.6 by specifying:

>     Requires-Python: >=3.6

Reviewer: Oldest non-EOL

```suggestion
    Requires-Python: >=3.8
```

.. versionadded:: 1.2

This field specifies the Python version(s) that the distribution is

> compatible with. Installation tools may look at this when

Reviewer 2: The way I think about this field (and the dependency version constraints) is something along the lines of "let's exclude the versions that we know we are not compatible with at the time of writing". But I do not know how to phrase this properly.

Author: Agreed - the "exclude incompatible" use case is elaborated somewhat in the (new) example offered below.

Reviewer 3: How about something along the lines of "A typical use case for this field is allowing the packagers to declare Python versions that are known to be incompatible with the distribution, which lets the installers disregard such distribution packages during the dependency resolution under incompatible runtimes".

picking which version of a project to install.

The value must be in the format specified in :doc:`version-specifiers`.

> For example, if a distribution uses `f-strings <https://docs.python.org/3/whatsnew/3.6.html#pep-498-formatted-string-literals>`_

Reviewer 3: Please, use intersphinx-provided references to link other Sphinx sites:
```suggestion
For example, if a distribution uses :ref:`f-strings <whatsnew36-pep498>`
```

> then it may prevent installation on Python < 3.6 by specifying:

Reviewer 3: Code blocks w/o an explicit directive require trailing double colon:
```suggestion
then it may prevent installation on Python < 3.6 by specifying::
```

For example, if a distribution uses `f-strings <https://docs.python.org/3/whatsnew/3.6.html#pep-498-formatted-string-literals>`_
then it may prevent installation on Python < 3.6 by specifying:

Examples::
    Requires-Python: >=3.6

    Requires-Python: >=3
    Requires-Python: >2.6,!=3.0.*,!=3.1.*

>     Requires-Python: ~=2.6

Author: Disagree on these. Excluding versions from range and compatible release are well covered in PEP 440, which is referenced several times in this document already.

Neither of these are typical use-cases for Requires-Python, but more edge cases.  `Requires-Python >2.6,!=3.0.*,!=3.1.*` was useful when projects still needed to support 2.7 + recent 3.x, but those Python versions are EOL now.

The compatible release ~= includes an upper bound, and should not normally be used with Requires-Python.

>     Requires-Python: ~=2.6

Reviewer 3: I disagree. By removing such example, we'd make them tribal knowledge. The PR description implies improvements and not getting rid of examples that make possibilities clear.

>     Requires-Python: ~=2.6

Author: Simplicity and conciseness is the goal, not detailing every possibility. Further, I replaced with a single example for a good reason - offering multiple examples could suggest that this field may be specified multiple times, which it may not. If you look on the [existing page](https://packaging.python.org/en/latest/specifications/core-metadata/), all the single use fields offer a single example. Apart from **Requires-Python**, only the fields marked as "(multiple use)" display multiple examples within the same code block.

>     Requires-Python: ~=2.6

Reviewer 4: @webknjaz I'm gonna move forward with this PR for now, since it is an overall improvement over status quo -- if you feel strongly about this, let's do a follow-up issue/PR to add a Python 2.7+ & 3.6+ example here.

> 

Reviewer 5: I will go two steps further now and ask if you can't just add more.  To be possibly heretic, something in the direction of:
```
.. note::
    At this time, the PyPA strongly suggests to never add upper version limits on this field, at least for libraries.
    For most packages an upper version pin is not necessary since even a Python 4.0 relase is
    not likely to break them.
    Even for libraries where an upper version pin is clearly correct, we strongly discourage adding the pin because:
    * It is often not useful in practice because current tooling means it is typically
      only useful if old versions support a wider range of Python versions.
    * Some tools like poetry and PDM force the bounds of dependencies on the package.
      That may be correct for lockfiles but is incorrect for this field for libraries:
      a library will not stop working with a new Python version because it has a
      dependency that does not *yet* work with a new Python version.
      Even for lockfiles the existing ecosystem tooling unfortunately may also lead
      to the first issue creating nuisance for users.

    Unfortunately, these issues need to be addressed in the ecosystem and by improving the
    tools before we can recommend upper version constraints. 
```
(even this is probably nonsense in the detail, and I don't know how much poetry/PDM are to blame exactly, although I doubt they are blameless.)
